src/hostname/hostnamed.c

```diff
@@ -367,33 +367,53 @@ static int context_update_kernel_hostname(
         return r; /* 0 if no change, 1 if something was done  */
 }
 
+static void unset_statp(struct stat **p) {
+        if (!*p)
+                return;
+
+        **p = (struct stat) {};
+}
+
 static int context_write_data_static_hostname(Context *c) {
+        _cleanup_(unset_statp) struct stat *s = NULL;
```

Comment:
maybe add a brief comment here explaining in short words what this does. this is not obvious at all i think. i.e. maybe "make sure that if we fail here, we invalidate the cached information, since it was updated already, even if we can't make it hit the disk" or so